nemo/collections/asr/parts/submodules/fast_rnnt_greedy_decoding.py

```diff
@@ -0,0 +1,440 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nemo.core.utils.cuda_python_utils import check_cuda_python_cuda_graphs_conditional_nodes_supported
+
+check_cuda_python_cuda_graphs_conditional_nodes_supported()
+
+import contextlib
+import ctypes
+import time
+from dataclasses import dataclass, field
+from itertools import product
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+from cuda import cuda, cudart, nvrtc
+from omegaconf import DictConfig, OmegaConf
+
+from nemo.collections.asr.modules import rnnt_abstract
+from nemo.collections.asr.parts.utils import rnnt_utils
+from nemo.collections.asr.parts.utils.asr_confidence_utils import ConfidenceMethodConfig, ConfidenceMethodMixin
+from nemo.collections.common.parts.rnn import label_collate
+from nemo.core.classes import Typing, typecheck
+from nemo.core.neural_types import AcousticEncodedRepresentation, ElementType, HypothesisType, LengthsType, NeuralType
+from nemo.utils import logging
+
+
+def ASSERT_DRV(err):
```

Comment:
Is this a universal approach, that can be reused in other code? If yes, is it worth moving this function and several functions below to a separate file?

Reply:
I agree with you. I moved cu_call and assert_drv to cuda_python_utils.py. I didn't move the run_nvrtc() function at this time. It is not fully general (for example, it allows extracting only one compiled cuda kernel from a compiled module, but there is no reason why we shouldn't be able to extract multiple).